### PR TITLE
[MRG][feature] Change warns signature to mimic the raises call

### LINKED ISSUE #2708: RFC: provide warns with a behavior similar to raises
When dealing with the messages thrown by warnings (IMHO) it would great if we could use the same calls as for the `raises` case.  As example, this would allow for: 

```py
    with warns(RuntimeWarning, message='my runtime warning'):
        warnings.warn("my runtime warning", RuntimeWarning)

    with warns(UserWarning, match=r'must be \d+$'):
        warnings.warn("value must be 42", UserWarning)
```
I have implemented such behavior by wrapping `pytest.warns` [here](https://github.com/scikit-learn-contrib/imbalanced-learn/pull/321/commits/8da92de17021ab89a9a12cd120d63819a7a721fe#diff-0cfeb0ca6012e38ec74597e8364be29aR130) in the following manner: 

```py
from pytest import warns as _warns
from contextlib import contextmanager
from re import compile

@contextmanager
def warns(expected_warning, message=None, match=None):
    with _warns(expected_warning) as record:
        yield

    wrn_msg = str(record[0].message)
    if message is not None:
        assert_msg = '"{}" is different from "{}"'.format(wrn_msg, message)
        assert wrn_msg == message, assert_msg
    elif match is not None:
        regex = compile(match)
        assert_msg = '"{}" pattern not found in "{}"'.format(match, wrn_msg)
        assert regex.search(wrn_msg) is not None, assert_msg
    else:
        pass
```

Do you think is a valuable feature? Do you think that a PR implementing such behavior is useful to anyone? I'm sure I'm not the first one to wander why `warns` has different signature than `raises`. Therefore I can only guess that is not that easy to integrate it to pytest. But If you think that PR is in order I'll be happy to do it.

### PR DESCRIPTION
**Pull request by massich**
fixes #2708 providing warns with `message` and `match` parameters (similarly to `pytest.raises`. see #2227)

### CODE REVIEW

**Comment by massich** [comment]
@nicoddemus Before diving into the code, I would like to get some feedback with the functionality itself. Could you check the tests I've written and ping to the appropriated developers to provide feedback before writing the actual feature. Maybe you guys have different view of the problem (like what you pointed in #2708)

File: testing/test_recwarn.py [hunk 1]
```diff
@@ -284,3 +284,27 @@ def test(run):
         ''')
         result = testdir.runpytest()
         result.stdout.fnmatch_lines(['*2 passed in*'])
+
+    def test_message_using(self):
+        source = "warnings.warn('my runtime warning', RuntimeWarning)"
+        pytest.warns(RuntimeWarning, source, message='my runtime warning')
```
Comment (RonnyPfannschmidt) [inline testing/test_recwarn.py]: warns should not replicate the old call style , and even more so for the execute code snippets style, its just an accident of history we should stay clear of

Comment (massich) [inline testing/test_recwarn.py]: so, warnings should always be used within a context manager?

Comment (massich) [inline testing/test_recwarn.py]: I've changed `test_message_using` in favor of `test_message`. (The using was a copy-paste error)

**Comment by nicoddemus** [comment]
Hi @massich sorry for not responding earlier, this one fell through the cracks.

The tests look good to me. The question of whatever we should match a single warning or error out vs match any of the warnings, I think the second is more practical so I would go for it.

I definitely agree we should support only the context manager case and treat function-form and (specially) string-eval forms as history.

cc @flub @hpk42 @The-Compiler would you guys like to share your opinions here?

**Comment by The-Compiler** [comment]
I don't really have an opinion on this, FWIW - I treat all warnings as errors :wink:

**Comment by massich** [comment]
I got `test_match_regex` working and I broke some other test, but these could be adressed afterwards. Could you give me some feedback regarding the functionality implementation. cc:@nicoddemus

**Comment by flub** [comment]
Hi, tests seem fine but I do t see any for the message kwarg. Should that not have been included according to the description?

**Comment by massich** [comment]
@flub [here](https://github.com/pytest-dev/pytest/pull/2711/files#diff-69c460e0b158e3d6097167dd7b6ecdfbR289) `warns` is called with `match=r'must be \d+$'`

File: testing/test_recwarn.py [hunk 2]
```diff
@@ -284,3 +284,23 @@ def test(run):
         ''')
         result = testdir.runpytest()
         result.stdout.fnmatch_lines(['*2 passed in*'])
+
+    def test_match_regex(self):
+        with pytest.warns(UserWarning, match=r'must be \d+$'):
+            warnings.warn("value must be 42", UserWarning)
+
+        with pytest.raises(pytest.fail.Exception):
```
Comment (nicoddemus) [inline testing/test_recwarn.py]: Please include a case which matches with `match` but the warning raised is wrong:

```python
with pytest.raises(pytest.fail.Exception):
   with pytest.warns(UserWarning, match=r'must be \d+$'):
       warnings.warn("value must be 42", RuntimeWarning)
```

File: _pytest/recwarn.py [hunk 3]
```diff
@@ -203,3 +208,15 @@ def __exit__(self, *exc_info):
                          "The list of emitted warnings is: {1}.".format(
                              self.expected_warning,
                              [each.message for each in self]))
+                elif self.match_expr is not None:
+                    for r in self:
+                        if issubclass(r.category, self.expected_warning):
+                            if compile(self.match_expr).search(str(r.message)):
```
Comment (nicoddemus) [inline _pytest/recwarn.py]: This is minor, but you are shadowing the builtin `compile`. I suggest changing to `re.compile` and `import re` instead.

**Comment by massich** [comment]
The changelog of the feature branch is outdated. Souldn't it have a 3.3.0 section? 
Could you update it so that I can rebase feature, or point out where should we state this feature in the changelog

**Comment by RonnyPfannschmidt** [comment]
@massich we use towncrier to stop the merge pain - please add a news fragment file in the changelog folder

**Comment by massich** [comment]
Woo thats cool. Thanks.

File: testing/test_recwarn.py (hunk 2 above)
Comment (massich) [reply testing/test_recwarn.py]: done.

**Comment by RonnyPfannschmidt** [comment]
well done, thanks 👍

**Pull request merged**